docker/bookkeeper/entrypoint.sh

```diff
@@ -48,31 +56,143 @@ export BK_tlsTrustStoreType=JKS
+   #
+   # If the `common.sh` file is sourced in a script that sets "exit on error" as true (set -e), the script shall return
+   # an error code `255`. Since, the `bookkeeper/bin/bookeeper` shell script sources the `common.sh` file, and it sets
+   # "exit on error" as true, we encounter this error when it is invoked. To avoid that error, here we modify the
+   # `common.sh` file to execute `/sbin/sysctl -n net.ipv6.bindv6only` only if file `/proc/sys/net/ipv6/bindv6only`
+   # is available.
+
+   # Replace the 22nd line with "if [ -f /sbin/sysctl ] && [ -f /proc/sys/net/ipv6/bindv6only ]; then".
+   sed -i "22s|.*|if [ -f /sbin/sysctl ] \&\& [ -f /proc/sys/net/ipv6/bindv6only ]; then|" ${BK_HOME}/bin/common.sh
+}
+
+initialize_cluster() {
+    set +e
+
+    zk-shell --run-once "ls ${BK_zkLedgersRootPath}/available/readonly" ${BK_zkServers}
+    if [ $? -eq 0 ]; then
+        echo "Cluster metadata already exists"
+    else
+        # Create an ephemeral zk node `bkInitLock` for use as a lock.
+        lock=`zk-shell --run-once "create ${BK_CLUSTER_ROOT_PATH}/bkInitLock '' true false false" ${BK_zkServers}`
+        if [ -z "$lock" ]; then
+            echo "Bookkeeper znodes do not exist in Zookeeper. Initializing a new Bookeekeper cluster."
+
+            # Note that this `bookkeeper` shell script sets "exit on error" (`set -e`) and sources from
+            # ${SCRIPTS_DIR}/common.sh. Make sure to invoke `fix_bk_ipv6_check()` before the control reaches here.
+            /opt/bookkeeper/bin/bookkeeper shell initnewcluster
+            if [ $? -eq 0 ]; then
+                echo "initnewcluster operation succeeded"
+            else
+                echo "initnewcluster operation failed. Please check the reason."
+                echo "Exit status of initnewcluster"
+                echo $?
+                exit
+            fi
+        else
+            echo "Others may be initializing the cluster at the same time."
+            tenSeconds=1
+            while [ ${tenSeconds} -lt 100 ]
+            do
+                sleep 10
+                zk-shell --run-once "ls ${BK_zkLedgersRootPath}/available/readonly" ${BK_zkServers}
+                if [ $? -eq 0 ]; then
+                    echo "Waited $tenSeconds * 10 seconds. Successfully listed ''${BK_zkLedgersRootPath}/available/readonly'"
+                    break
+                else
+                    echo "Waited $tenSeconds * 10 seconds. Continue waiting."
+                    (( tenSeconds++ ))
+                    continue
+                fi
+            done
+
+            if [ ${tenSeconds} -eq 100 ]; then
+                echo "Waited 100 seconds for bookkeeper cluster to initialize, but to no avail. Something is wrong, please check."
+                exit
+            fi
+        fi
+    fi
+    set -e
+}
+
+format_bookie_data_and_metadata() {
+    if [ `find $BK_journalDirectory $BK_ledgerDirectories $BK_indexDirectories -type f 2> /dev/null | wc -l` -gt 0 ]; then
```

Comment:
The comments in this block say "probably" a few times. It doesn't give a lot of confidence to the action being taken. Is there anything we can do to reduce the uncertainty?

We need to at least create issues to track these cases, assuming there are corner cases that can be problematic.

Reply:
These are an existing piece of code and comments which were refactored into a function for modularity.

Reply:
Sorry for the late review, I remember this issue can be traced back several months ago when Adrian found that if we don't format bookie when bookie starts, the system will have InvalidCookieException if there are some metadata of the previous deployment left in the zookkeeper. However, we don't want to format the bookie when the bookkeeper container is just restarted or upgraded, otherwise it will cause data loss.

The main purpose here is basically to use `bookieformat -deleteCookie ` before a new deployment to prevent InvalidCookieException.

Reply:
Thanks for the clarification, @Tristan1900. @fpj I've updated the comments in the latest commit. Please see if that resolves your concern here.